binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+We first write some random numbers to a parquet file:
+
+```{r parquet-precision-write}
+#| message: false
+parquet_file <- tempfile(fileext = ".parquet")
+original <- tibble(old_x = rnorm(1000))
+write_parquet(original, parquet_file)
+```
+
+Next we load the file and give the reloaded variable a new name:
+
+```{r parquet-precision-read}
+reloaded <- read_parquet(parquet_file) |>
+  rename(new_x = old_x)
+```
+
+Finally we bind the old and new data sets together and compare them:
+
+```{r parquet-precision-success}
+bind_cols(original, reloaded) |> 
+  mutate(same_value = old_x == new_x) |>
+  count(same_value)
+```
+
+As you can see, numeric values stored in parquet files are stored faithfully.
+Loading the data from file produces the same values in the original data set.
+
+### Parquet files are smaller
+
+We can use the Seattle library data set to highlight the advantage in file size.
+The `write_parquet()` function can take an Arrow Dataset object like `seattle_lib` as the input, so we can write the file like this:
+
+```{r write-parquet}
+#| eval: !expr "!file.exists('data/seattle-library-checkouts.parquet')"
+write_parquet(seattle_lib, "data/seattle-library-checkouts.parquet")
+```
+
+Now let's compare the two files in size:
+
+```{r compare-file-sizes}
+file.size("data/seattle-library-checkouts.csv")
+file.size("data/seattle-library-checkouts.parquet")
+```
+
+The parquet file is about half the size of the csv file.
+
+### Parquet files are faster
+
+The format of your data file can have a substantial impact on how long a data analysis pipeline takes to execute.
+To see this, let's time how long it takes to count the number of books checked out in each month of 2021 when the data are stored in csv format:
+
+```{r dataset-performance-csv}
+#| cache: true
+system.time({
+  open_dataset(
+    sources = "data/seattle-library-checkouts.csv", 
+    format = "csv"
+  ) |>
+    filter(CheckoutYear == 2021, MaterialType == "BOOK") |>
+    group_by(CheckoutMonth) |>
+    summarise(TotalCheckouts = sum(Checkouts)) |>
+    arrange(desc(CheckoutMonth)) |>
+    collect()
+})
+```
+
+Now let's repeat the exercise using the parquet file:
+
+```{r dataset-performance-parquet}
+#| cache: true
+system.time({
+  open_dataset(
+    sources = "data/seattle-library-checkouts.parquet", 
+    format = "parquet"
+  ) |>
+    filter(CheckoutYear == 2021, MaterialType == "BOOK") |>
+    group_by(CheckoutMonth) |>
+    summarise(TotalCheckouts = sum(Checkouts)) |>
+    arrange(desc(CheckoutMonth)) |>
+    collect()
+})
+```
+
+The analysis of the parquet data finishes in about one tenth the time it takes to perform the same analysis with the csv data.
+
+## Multi-file data sets
+
+When data sets become large it is often useful to split the data across many files rather than storing everything in a single large file.
+When this structuring is done intelligently, the multi-file strategy can lead to significant improvements in performance.
```

Comment:
Because it allows multiple cores to work simultaneously?

Reply:
`arrow` does multithreaded reading with single files too

Reply:
yeah I was thinking about partitioning. I've reworded to hint to the reader that the advantage comes from reading only a subset of files. Trying to be slightly vague here because it doesn't exactly feel like the spot to start talking about predicate pushdown!